### ISSUE TITLE
Build environment? Cannot build on ubuntu.

### ISSUE THREAD

**Issue by davidfiala** [deeplay-io/nice-grpc#590]
Hi,

I was hoping to take a look at https://github.com/deeplay-io/nice-grpc/issues/555 but I noticed that on a fresh Ubuntu 23.10 with node v20.13.1 installed freshly via nvm I was unable to get the project to build properly without lots of hacking. And even then, I cannot get tests working.

What is your current developmental workstation build setup? Does it work with a fresh version of node? (ie, does it work if you haven't done any global npm installs?)

Some immediate observations:

- `@types/node` seems needed, but isn't present in the main package.json
- `packageManager` is missing in package.json, but I do see a `yarn.lock` file

I'd be happy to help contribute here (including starting with a `BUILDING.md`), but would appreciate some pointers to at least see what a working baseline build environment looks like. Thank you!

**Comment by aikoven** [comment]
Hey @davidfiala!

I am not directly opposed to your modernization proposal, but are you sure all of this is required to fix build problems? What problem exactly you got with the build by the way?

Below I will leave comments on each of the proposals. But something tells me that we can unlock your work on #555 with much much less blood.

> My other goal would be to get tsconfig aligned across the project and get vscode to play ball with the tsconfig. Right now, I think some of the src/file expressions preclude some files from properly being recognized as in the project or not, leading to red squiggles and inability to perform intellisense.

What files are we talking about? I am using VSCode to develop nice-grpc and everything works just fine for me.

> upgrade over to yarn 4.x

That would be nice! I would also consider migrating to npm, to simplify things. Using yarn is just my old-school habit.

> switch to modern stephenh/ts-proto for generation and drop google-protobuf (not seeing much action these days)

The only package that uses google-protobuf in its implementation currently is the server reflection library. That's because reflection protobufs are written in protobuf v2, which at that time was supported by ts-proto only for parsing but not serialization. It would be great to migrate if that's possible now.

Apart from that, we provide _support_ for google-protobuf in nice-grpc and nice-grpc-web and I don't want to drop it, because it can aid the migration of old code to nice-grpc from grpc-js/grpc-web where google-protobuf is still often used.

> upgrade lib/target towards ES2018 or higher

It's fine for NodeJS, but most of our packages can be used in the browser and other platforms. That's why we use [@tsconfig/recommended](https://www.npmjs.com/package/@tsconfig/recommended), which currently has es2016 target. I worry that this change could break someone.

> switch to ESM internally everywhere

> I think we can still output in any format if we had to. esbuild, etc could help if tsc gets annoying.

I have some concerns about that.

First of all, I don't see a way we could drop CommonJS output for now. This would be more or less acceptable if our only consumer was application code. But our packages are broadly used in libraries, and this change would force them to switch to ESM as well.

Having dual output sounds better, but has a price of a more complicated build system. To be honest, I don't see benefits that would justify this change.

---

To sum up, I am definitely up for modernizing things, especially if that leads to simplification. On the other hand, complicating things does not sound right to me, even if that leads to a more modern setup.

That said, I sure hope that you acknowledge that all of the above is largely unrelated to #555.

> I'd be happy to jump on a chat anytime to discuss realtime, too. WDYT?

Absolutely! I'm on Telegram [@aikoven](https://t.me/aikoven) and Discord [@aikoven](https://discordapp.com/users/120932198676299777)

**Comment by aikoven** [comment]
Thanks for the reproduction steps!

I see now that the problem is caused by yarn 4 producing a little different `node_modules` layout (particularly the location of binaries).

Here are the versions installed on my Mac:

```bash
❯ node --version
v18.20.2

❯ yarn --version
1.22.22
```

I checked the CI logs and it appears that GitHub `setup-node` action also uses yarn `1.22.22` on both [NodeJS 18](https://github.com/deeplay-io/nice-grpc/actions/runs/9346758191/job/25722283416#step:3:12) and [NodeJS 20](https://github.com/deeplay-io/nice-grpc/actions/runs/9346758191/job/25722283904#step:3:12).

I'm not sure whether it is possible to lock yarn to v1 via `packageManager` field. Can you please try the setup with `"packageManager": "yarn@1.22.22"`?

I also tried installing dependencies using `npm`, which resulted in same errors with missing binaries. If the `packageManager` setting won't help, we could try to update the binary paths to match the structure produced by `npm`.
